### PR TITLE
feat: CAIP Multichain (OLD ignore. Only for reference now)

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**

~~This branch is becoming the feature branch for CAIP-25, CAIP-27, CAIP stream bifurcation, and CAIP-25 Permissions~~

This is branch is now stale.

develop (and [core package](https://github.com/MetaMask/core/pull/4784)) -> [CAIP-25 permission migration (uses new core package)](https://github.com/MetaMask/metamask-extension/pull/27847) -> [New multichain branch](https://github.com/MetaMask/metamask-extension/pull/27782)

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24181?quickstart=1)

## **Related issues**

See: https://github.com/MetaMask/MetaMask-planning/issues/2360
See: https://github.com/MetaMask/test-dapp/pull/324

## **Manual testing steps**

1. Go to this page...
2.
3.

```
BARAD_DUR=1 CHAIN_PERMISSIONS=1 yarn start
```

**Replace EXTENSION_ID with your local extension ID**
```
const EXTENSION_ID = 'nonfpcflonapegmnfeafnddgdniflbnk';
const extensionPort = chrome.runtime.connect(EXTENSION_ID)
extensionPort.onMessage.addListener((msg) => console.log('extensionPort on message', msg))
extensionPort.postMessage({
    type: 'caip-x',
    data: {
        "jsonrpc": "2.0",
        method: 'wallet_createSession',
        params: {
            requiredScopes: {
                'eip155': {
                    references: ['1', '59144'],
                    methods: [
                        'eth_sendTransaction',
                        'eth_accounts',
                        'eth_blockNumber',
                        'eth_getBalance',
                        'personal_sign',
                        'eth_requestAccounts',
                        'wallet_getPermissions',
                        'wallet_requestPermissions',
                        'wallet_revokePermissions',
                        'eth_chainId',
                        'wallet_switchEthereumChain'
                    ],
                    notifications: ['accountsChanged', 'chainChanged'],
                },
                'eip155:11155111': {
                    methods: ['eth_blockNumber'],
                    notifications: ['accountsChanged', 'chainChanged'],
                },
                'wallet': {
                    methods: [
                        'wallet_getPermissions',
                    ],
                    notifications: [],
                },
            },
            optionalScopes: {
                'eip155:64': {
                    methods: [
                        'eth_blockNumber',
                        'eth_getBalance',
                        'eth_getBlockByHash',
                        'eth_getBlockByNumber',
                    ],
                    notifications: ['accountsChanged', 'chainChanged'],
                },
            },
            sessionProperties: {
                expiry: 1,
                'caip154-mandatory': 'true',
            },
        },
    }
})
extensionPort.postMessage({
    type: 'caip-x',
    data: {
        "jsonrpc": "2.0",
        "method": "wallet_invokeMethod",
        "params": {
            "sessionId": "0xdeadbeef",
            "scope": "eip155:1",
            "request": {
                "method": "eth_chainId",
            }
        }
    }
})
extensionPort.postMessage({
    type: 'caip-x',
    data: {
        "jsonrpc": "2.0",
        "method": "wallet_invokeMethod",
        "params": {
            "sessionId": "0xdeadbeef",
            "scope": "eip155:11155111",
            "request": {
                "method": "eth_chainId",
            }
        }
    }
})
```

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by jiexi** [comment]
@metamaskbot update-policies

**Comment by adonesky1** [comment]
@metamaskbot update-policies

File: app/scripts/lib/multichain-api/provider-authorize.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+// TODO:
+// When user disapproves accepting calls with the request methods
+//   code = 5001
+//   message = "User disapproved requested methods"
+// When user disapproves accepting calls with the request notifications
+//   code = 5002
+//   message = "User disapproved requested notifications"
+
+export async function providerAuthorizeHandler(req, res, _next, end, hooks) {
+  // TODO: Does this handler need a rate limiter/lock like the one in eth_requestAccounts?
```
Comment (adonesky1) [inline app/scripts/lib/multichain-api/provider-authorize.js:36]: no it has it by default with the approval controller unless explicitly specified as not needing rate limiting

File: app/scripts/lib/multichain-api/provider-authorize.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    );
+  }
+
+  const sessionId = '0xdeadbeef';
+
+  if (sessionProperties && Object.keys(sessionProperties).length === 0) {
+    return end(
+      new EthereumRpcError(5300, 'Invalid Session Properties requested'),
+    );
+  }
```
Comment (adonesky1) [inline app/scripts/lib/multichain-api/provider-authorize.js:65]: This is probably something else we don't need to blow up the whole connection for

**Comment by adonesky1** [comment]
@metamaskbot update-policies

File: app/scripts/lib/multichain-api/provider-authorize/handler.js [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+// TODO:
+// When user disapproves accepting calls with the request methods
+//   code = 5001
+//   message = "User disapproved requested methods"
+// When user disapproves accepting calls with the request notifications
+//   code = 5002
+//   message = "User disapproved requested notifications"
+
+export async function providerAuthorizeHandler(req, res, _next, end, hooks) {
+  // TODO: Does this handler need a rate limiter/lock like the one in eth_requestAccounts?
```
Comment (adonesky1) [inline app/scripts/lib/multichain-api/provider-authorize/handler.js:43]: No. It is rate limited by default

**Comment by jiexi** [comment]
@metamaskbot update-policies

File: .circleci/config.yml [hunk 4]
```diff
@@ -870,7 +873,7 @@ jobs:
           at: .
       - run:
           name: Build extension for testing
-          command: yarn build:test
+          command: CHAIN_PERMISSIONS=1 BARAD_DUR=1 yarn build:test
```
Comment (jiexi) [inline .circleci/config.yml:876]: this needs to move into it's own build step

File: html-report-caip27/index.html
Comment (jiexi) [inline html-report-caip27/index.html:1]: need to remove this

**Comment by jiexi** [comment]
@metamaskbot update-policies

File: app/scripts/metamask-controller.js [hunk 5]
```diff
@@ -1540,7 +1534,7 @@ export default class MetamaskController extends EventEmitter {
         },
       },
       env: {
-        isAccountSyncingEnabled: isManifestV3,
+        isAccountSyncingEnabled: false, // TODO: undo this once fixed
```
Comment (jiexi) [inline app/scripts/metamask-controller.js:1537]: TODO

**Comment by jiexi** [comment]
This branch is now stale

**Comment by jiexi** [comment]
All comments have been copied out. This PR can be ignored now

**Pull request closed without merging**